Relax equality check

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165460



When an object is inherited from multiple types, the previous check would fail. So we should relax it to respect eager semantic 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Differential Revision: [D84635322](https://our.internmc.facebook.com/intern/diff/D84635322)

tugsbayasgalan committed:
Relax equality check

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Relax equality check"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Relax equality check"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

avikchaudhuri reviewed (approved):
Still not entirely satisfied whether `a == b` and `b == a` will be `True` / `False` in exactly the same cases when `type(a)` is a subclass of `type(b)` or `type(b)` is a subclass of `type(a)` but don't want to block.

tugsbayasgalan committed:
Update on "Relax equality check"




When an object is inherited from multiple types, the previous check would fail. So we should relax it to respect eager semantic 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D84635322](https://our.internmc.facebook.com/intern/diff/D84635322)

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -197,9 +197,11 @@ def __hash__(self):
             # TODO: Put this in utils and share it between variables/builtin.py and here
-            if type(a) is not type(b):
+            type_a, type_b = type(a), type(b)
+            if not (issubclass(type_a, type_b) or issubclass(type_b, type_a)):
```
Agree with this .. this patch should be enough

```
diff --git a/torch/_dynamo/variables/dicts.py b/torch/_dynamo/variables/dicts.py
index 3379206d81b..e13137a5406 100644
--- a/torch/_dynamo/variables/dicts.py
+++ b/torch/_dynamo/variables/dicts.py
@@ -197,7 +197,11 @@ class ConstDictVariable(VariableTracker):
         @staticmethod
         def _eq_impl(a, b):
             # TODO: Put this in utils and share it between variables/builtin.py and here
-            if type(a) is not type(b):
+            if isinstance(a, type(b)):
+                return a.__eq__(b)
+            elif isinstance(b, type(a)):
+                return b.__eq__(a)
+            elif type(a) is not type(b):
                 return False
             elif isinstance(a, tuple):
                 Hashable = ConstDictVariable._HashableTracker
```

tugsbayasgalan replied:
Hmm the reason we have it this way is that the signature of this function checks if a == b, so i feel even if there is assymmetry, we should respect that. I think it would have failed in eager anyways.

anijain2305 commented on test/export/test_export.py:
```diff
@@ -15180,6 +15181,45 @@ def forward(self, x, y):
                 filtered_nn_module_stack[1], "mod_list_2.slice(4, 5, None).0"
             )
 
+    def test_enum_str(self):
```
Please move the test case to test/dynamo/test_dicts.py. Test fragmentation causes pain because some tests in random folder fail while we think everything should be covered in test_dicts.py.

Try torch.compile here instead of fullgraph capture, when you move the test to `test_dicts.py`.

tugsbayasgalan committed:
Update on "Relax equality check"




When an object is inherited from multiple types, the previous check would fail. So we should relax it to respect eager semantic 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D84635322](https://our.internmc.facebook.com/intern/diff/D84635322)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Relax equality check"




When an object is inherited from multiple types, the previous check would fail. So we should relax it to respect eager semantic 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D84635322](https://our.internmc.facebook.com/intern/diff/D84635322)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D84635322).
